src/stream/src/executor/aggregation/minput.rs

```diff
@@ -43,61 +74,1182 @@ impl<S: StateStore> MaterializedInputState<S> {
+                ),
+                AggKind::StringAgg => {
+                    Box::new(GenericStateCache::new(StringAgg, cache_capacity, row_count))
+                }
+                AggKind::ArrayAgg => {
+                    Box::new(GenericStateCache::new(ArrayAgg, cache_capacity, row_count))
+                }
                 _ => panic!(
                     "Agg kind `{}` is not expected to have materialized input state",
                     agg_call.kind
                 ),
-            },
+            };
+
+        Self {
+            group_key: group_key.cloned(),
+            arg_col_indices,
+            state_table_arg_col_indices,
+            order_col_indices,
+            state_table_order_col_indices,
+            cache,
+            cache_key_serializer,
+            _phantom_data: PhantomData,
         }
     }
 
-    /// Apply a chunk of data to the state.
-    pub async fn apply_chunk(
+    /// Apply a chunk of data to the state cache.
+    pub fn apply_chunk(
         &mut self,
         ops: Ops<'_>,
         visibility: Option<&Bitmap>,
         columns: &[&ArrayImpl],
-        state_table: &mut StateTable<S>,
     ) -> StreamExecutorResult<()> {
-        self.inner
-            .apply_chunk(ops, visibility, columns, state_table)
-            .await
+        self.cache.apply_batch(StateCacheInputBatch::new(
+            ops,
+            visibility,
+            columns,
+            &self.cache_key_serializer,
+            &self.arg_col_indices,
+            &self.order_col_indices,
+        ));
+        Ok(())
     }
 
     /// Get the output of the state.
     pub async fn get_output(&mut self, state_table: &StateTable<S>) -> StreamExecutorResult<Datum> {
-        self.inner.get_output(state_table).await
+        if !self.cache.is_synced() {
+            let all_data_iter = iter_state_table(state_table, self.group_key.as_ref()).await?;
+            pin_mut!(all_data_iter);
+
+            let mut cache_filler = self.cache.begin_syncing();
+            #[for_await]
+            for state_row in all_data_iter.take(cache_filler.capacity()) {
+                let state_row = state_row?;
+                let cache_key = {
+                    let mut cache_key = Vec::new();
+                    self.cache_key_serializer.serialize_datums(
+                        self.state_table_order_col_indices
+                            .iter()
+                            .map(|col_idx| &(state_row.0)[*col_idx]),
+                        &mut cache_key,
+                    );
+                    cache_key
+                };
+                let cache_value = self
+                    .state_table_arg_col_indices
+                    .iter()
+                    .map(|i| state_row[*i].as_ref().map(ScalarImpl::as_scalar_ref_impl))
+                    .collect();
+                cache_filler.insert(cache_key, cache_value);
+            }
+        }
+        debug_assert!(self.cache.is_synced());
```

Comment:
We can enable the assertion in production as it's lightweight. :)

```suggestion
        assert!(self.cache.is_synced());
```